File: test/worker_test.dart
library emscripten.test.worker;

import 'package:test/test.dart';
import 'package:emscripten/emscripten.dart';

testWorker() {
  group('worker', () {
    AsyncWorker worker;
    setUp(() {
      worker = new AsyncWorker('worker_test.js');
    });
    test('post', () async {
      var message = ['sum', 2, 2];
      var result = await worker.post(message);
      expect(result, equals(4));
    });
    test('error', () {
      var message = ['error'];
      worker.post(message).then((result) {
        fail(result);
      }, onError: (err, st) {
        expect(err, 'error message');
      });
    });
  });
}

File: test/worker_test.js
if (this.document === undefined) {
  this.onmessage = function(event) {
    var payload = event["data"];
    var id = payload[0], message = payload[1];
    var cmd = message[0];
    var result;

    if (cmd == "sum") {
      result = message[1] + message[2];
    }

    if (cmd == "error") {
      result = ["error", "error message"];
    }

    this.postMessage([id, result]);
  }
}
